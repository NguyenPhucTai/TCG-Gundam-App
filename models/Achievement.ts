export interface Achievement {
  id: string;
  title: string;
  description: string;
  icon: string;
  condition: AchievementCondition;
  isUnlocked: boolean;
  unlockedAt?: Date;
}

export interface AchievementCondition {
  type: 'streak' | 'total_days' | 'habits_count' | 'completion_rate';
  target: number;
  habitId?: string; // Nếu achievement dành cho habit cụ thể
}

export interface UserAchievement {
  userId: string;
  achievementId: string;
  unlockedAt: Date;
  progress: number;
}
