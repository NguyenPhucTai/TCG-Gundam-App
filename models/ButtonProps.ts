import React from 'react';

// Interface for Button component props
export interface ButtonProps {
  children: React.ReactNode;
  onPress: () => void;
  variant?: 'primary' | 'secondary' | 'control';
  disabled?: boolean;
}
