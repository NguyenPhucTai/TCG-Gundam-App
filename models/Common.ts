export interface Category {
  id: string;
  name: string;
  description: string;
  icon: string;
  color: string;
}

export interface WeeklyReport {
  userId: string;
  weekStartDate: string;
  weekEndDate: string;
  totalHabits: number;
  completedDays: number;
  totalPossibleDays: number;
  completionRate: number;
  bestStreak: number;
  habitDetails: HabitWeeklyDetail[];
}

export interface HabitWeeklyDetail {
  habitId: string;
  habitName: string;
  completedDays: number;
  totalDays: number;
  streak: number;
}
