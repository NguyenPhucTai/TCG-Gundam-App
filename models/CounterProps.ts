// Interface for Counter component props
export interface CounterProps {
  label: string;
  value: number;
  hasButtons?: boolean;
  formatValue?: (value: number) => string;
  onIncrement?: () => void;
  onDecrement?: () => void;
}
