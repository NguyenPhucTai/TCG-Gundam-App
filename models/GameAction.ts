// Interface for game action history entries
export interface GameAction {
  id: string;
  timestamp: number;
  playerId: string;
  actionType: 'increment' | 'decrement' | 'reset';
  statType: 'resource' | 'exResource' | 'level';
  previousValue: number;
  newValue: number;
}
