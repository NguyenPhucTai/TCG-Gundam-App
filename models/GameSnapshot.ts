import { PlayerState } from './PlayerState';

// Snapshot of game state without history
export interface GameSnapshot {
  firstPlayer: PlayerState;
  secondPlayer: PlayerState;
}
