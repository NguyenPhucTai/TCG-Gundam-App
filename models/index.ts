// Export tất cả các interface từ models
export * from './Habit';
export * from './User';
export * from './Achievement';
export * from './Common';
