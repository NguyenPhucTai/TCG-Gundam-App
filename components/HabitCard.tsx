import React from 'react';
import { View, Text, StyleSheet, TouchableOpacity } from 'react-native';
import { Habit } from '../models';

interface HabitCardProps {
  habit: Habit;
  onToggle: (habitId: string) => void;
  onPress?: (habit: Habit) => void;
  testID?: string;
}

export const HabitCard: React.FC<HabitCardProps> = ({
  habit,
  onToggle,
  onPress,
  testID,
}) => {
  const handleToggle = () => {
    onToggle(habit.id);
  };

  const handlePress = () => {
    onPress?.(habit);
  };

  return (
    <TouchableOpacity
      style={[styles.container, { borderLeftColor: habit.color }]}
      onPress={handlePress}
      testID={testID}
      accessibilityRole="button"
      accessibilityLabel={`Habit: ${habit.name}`}
    >
      <View style={styles.content}>
        <Text style={styles.name}>{habit.name}</Text>
        {habit.description && (
          <Text style={styles.description}>{habit.description}</Text>
        )}
        <Text style={styles.streak}>
          Streak: {habit.streak} ngày
        </Text>
      </View>
      
      <TouchableOpacity
        style={[
          styles.checkButton,
          habit.isCompletedToday && styles.checkButtonCompleted,
        ]}
        onPress={handleToggle}
        testID={`${testID}-toggle`}
        accessibilityRole="checkbox"
        accessibilityState={{ checked: habit.isCompletedToday }}
      >
        <Text style={[
          styles.checkText,
          habit.isCompletedToday && styles.checkTextCompleted,
        ]}>
          {habit.isCompletedToday ? '✓' : '○'}
        </Text>
      </TouchableOpacity>
    </TouchableOpacity>
  );
};

const styles = StyleSheet.create({
  container: {
    backgroundColor: '#fff',
    borderRadius: 12,
    padding: 16,
    marginVertical: 8,
    marginHorizontal: 16,
    borderLeftWidth: 4,
    flexDirection: 'row',
    alignItems: 'center',
    shadowColor: '#000',
    shadowOffset: {
      width: 0,
      height: 2,
    },
    shadowOpacity: 0.1,
    shadowRadius: 3.84,
    elevation: 5,
  },
  content: {
    flex: 1,
  },
  name: {
    fontSize: 18,
    fontWeight: '600',
    color: '#333',
    marginBottom: 4,
  },
  description: {
    fontSize: 14,
    color: '#666',
    marginBottom: 8,
  },
  streak: {
    fontSize: 12,
    color: '#999',
    fontWeight: '500',
  },
  checkButton: {
    width: 40,
    height: 40,
    borderRadius: 20,
    borderWidth: 2,
    borderColor: '#E5E5EA',
    alignItems: 'center',
    justifyContent: 'center',
    backgroundColor: '#F9F9F9',
  },
  checkButtonCompleted: {
    backgroundColor: '#34C759',
    borderColor: '#34C759',
  },
  checkText: {
    fontSize: 20,
    color: '#999',
  },
  checkTextCompleted: {
    color: 'white',
    fontWeight: 'bold',
  },
});
