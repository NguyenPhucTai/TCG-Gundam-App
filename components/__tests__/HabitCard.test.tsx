import React from 'react';
import { render, fireEvent, screen } from '@testing-library/react-native';
import { HabitCard } from '../HabitCard';
import { Habit } from '../../models';

describe('HabitCard Component', () => {
  const mockOnToggle = jest.fn();
  const mockOnPress = jest.fn();

  const mockHabit: Habit = {
    id: '1',
    name: 'Đọc sách',
    description: 'Đọc ít nhất 30 phút mỗi ngày',
    streak: 5,
    isCompletedToday: false,
    color: '#FF6B6B',
  };

  beforeEach(() => {
    mockOnToggle.mockClear();
    mockOnPress.mockClear();
  });

  it('renders habit information correctly', () => {
    render(
      <HabitCard
        habit={mockHabit}
        onToggle={mockOnToggle}
        onPress={mockOnPress}
        testID="habit-card"
      />
    );

    expect(screen.getByText('Đọc sách')).toBeTruthy();
    expect(screen.getByText('Đọc ít nhất 30 phút mỗi ngày')).toBeTruthy();
    expect(screen.getByText('Streak: 5 ngày')).toBeTruthy();
  });

  it('calls onPress when card is pressed', () => {
    render(
      <HabitCard
        habit={mockHabit}
        onToggle={mockOnToggle}
        onPress={mockOnPress}
        testID="habit-card"
      />
    );

    const card = screen.getByTestId('habit-card');
    fireEvent.press(card);

    expect(mockOnPress).toHaveBeenCalledWith(mockHabit);
    expect(mockOnToggle).not.toHaveBeenCalled();
  });

  it('calls onToggle when check button is pressed', () => {
    render(
      <HabitCard
        habit={mockHabit}
        onToggle={mockOnToggle}
        onPress={mockOnPress}
        testID="habit-card"
      />
    );

    const toggleButton = screen.getByTestId('habit-card-toggle');
    fireEvent.press(toggleButton);

    expect(mockOnToggle).toHaveBeenCalledWith('1');
    expect(mockOnPress).not.toHaveBeenCalled();
  });

  it('displays uncompleted state correctly', () => {
    render(
      <HabitCard
        habit={mockHabit}
        onToggle={mockOnToggle}
        testID="habit-card"
      />
    );

    const toggleButton = screen.getByTestId('habit-card-toggle');
    expect(toggleButton.props.accessibilityState.checked).toBe(false);
    expect(screen.getByText('○')).toBeTruthy();
  });

  it('displays completed state correctly', () => {
    const completedHabit: Habit = {
      ...mockHabit,
      isCompletedToday: true,
    };

    render(
      <HabitCard
        habit={completedHabit}
        onToggle={mockOnToggle}
        testID="habit-card"
      />
    );

    const toggleButton = screen.getByTestId('habit-card-toggle');
    expect(toggleButton.props.accessibilityState.checked).toBe(true);
    expect(screen.getByText('✓')).toBeTruthy();
  });

  it('renders without description when not provided', () => {
    const habitWithoutDescription: Habit = {
      ...mockHabit,
    };
    delete (habitWithoutDescription as any).description;

    render(
      <HabitCard
        habit={habitWithoutDescription}
        onToggle={mockOnToggle}
        testID="habit-card"
      />
    );

    expect(screen.getByText('Đọc sách')).toBeTruthy();
    expect(screen.queryByText('Đọc ít nhất 30 phút mỗi ngày')).toBeNull();
    expect(screen.getByText('Streak: 5 ngày')).toBeTruthy();
  });

  it('works without onPress callback', () => {
    render(
      <HabitCard
        habit={mockHabit}
        onToggle={mockOnToggle}
        testID="habit-card"
      />
    );

    const card = screen.getByTestId('habit-card');
    
    // Should not throw error when pressed without onPress
    expect(() => fireEvent.press(card)).not.toThrow();
  });

  it('has correct accessibility properties', () => {
    render(
      <HabitCard
        habit={mockHabit}
        onToggle={mockOnToggle}
        testID="habit-card"
      />
    );

    const card = screen.getByTestId('habit-card');
    const toggleButton = screen.getByTestId('habit-card-toggle');

    expect(card.props.accessibilityRole).toBe('button');
    expect(card.props.accessibilityLabel).toBe('Habit: Đọc sách');
    expect(toggleButton.props.accessibilityRole).toBe('checkbox');
  });

  it('matches snapshot for uncompleted habit', () => {
    const tree = render(
      <HabitCard
        habit={mockHabit}
        onToggle={mockOnToggle}
        testID="habit-card"
      />
    );
    expect(tree.toJSON()).toMatchSnapshot();
  });

  it('matches snapshot for completed habit', () => {
    const completedHabit: Habit = {
      ...mockHabit,
      isCompletedToday: true,
    };

    const tree = render(
      <HabitCard
        habit={completedHabit}
        onToggle={mockOnToggle}
        testID="habit-card"
      />
    );
    expect(tree.toJSON()).toMatchSnapshot();
  });
});
