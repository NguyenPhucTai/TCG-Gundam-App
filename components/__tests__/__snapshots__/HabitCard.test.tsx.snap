// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`HabitCard Component matches snapshot for completed habit 1`] = `
<View
  accessibilityLabel="Habit: Đọc sách"
  accessibilityRole="button"
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": undefined,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    {
      "alignItems": "center",
      "backgroundColor": "#fff",
      "borderLeftColor": "#FF6B6B",
      "borderLeftWidth": 4,
      "borderRadius": 12,
      "elevation": 5,
      "flexDirection": "row",
      "marginHorizontal": 16,
      "marginVertical": 8,
      "opacity": 1,
      "padding": 16,
      "shadowColor": "#000",
      "shadowOffset": {
        "height": 2,
        "width": 0,
      },
      "shadowOpacity": 0.1,
      "shadowRadius": 3.84,
    }
  }
  testID="habit-card"
>
  <View
    style={
      {
        "flex": 1,
      }
    }
  >
    <Text
      style={
        {
          "color": "#333",
          "fontSize": 18,
          "fontWeight": "600",
          "marginBottom": 4,
        }
      }
    >
      Đọc sách
    </Text>
    <Text
      style={
        {
          "color": "#666",
          "fontSize": 14,
          "marginBottom": 8,
        }
      }
    >
      Đọc ít nhất 30 phút mỗi ngày
    </Text>
    <Text
      style={
        {
          "color": "#999",
          "fontSize": 12,
          "fontWeight": "500",
        }
      }
    >
      Streak: 
      5
       ngày
    </Text>
  </View>
  <View
    accessibilityRole="checkbox"
    accessibilityState={
      {
        "busy": undefined,
        "checked": true,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "alignItems": "center",
        "backgroundColor": "#34C759",
        "borderColor": "#34C759",
        "borderRadius": 20,
        "borderWidth": 2,
        "height": 40,
        "justifyContent": "center",
        "opacity": 1,
        "width": 40,
      }
    }
    testID="habit-card-toggle"
  >
    <Text
      style={
        [
          {
            "color": "#999",
            "fontSize": 20,
          },
          {
            "color": "white",
            "fontWeight": "bold",
          },
        ]
      }
    >
      ✓
    </Text>
  </View>
</View>
`;

exports[`HabitCard Component matches snapshot for uncompleted habit 1`] = `
<View
  accessibilityLabel="Habit: Đọc sách"
  accessibilityRole="button"
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": undefined,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    {
      "alignItems": "center",
      "backgroundColor": "#fff",
      "borderLeftColor": "#FF6B6B",
      "borderLeftWidth": 4,
      "borderRadius": 12,
      "elevation": 5,
      "flexDirection": "row",
      "marginHorizontal": 16,
      "marginVertical": 8,
      "opacity": 1,
      "padding": 16,
      "shadowColor": "#000",
      "shadowOffset": {
        "height": 2,
        "width": 0,
      },
      "shadowOpacity": 0.1,
      "shadowRadius": 3.84,
    }
  }
  testID="habit-card"
>
  <View
    style={
      {
        "flex": 1,
      }
    }
  >
    <Text
      style={
        {
          "color": "#333",
          "fontSize": 18,
          "fontWeight": "600",
          "marginBottom": 4,
        }
      }
    >
      Đọc sách
    </Text>
    <Text
      style={
        {
          "color": "#666",
          "fontSize": 14,
          "marginBottom": 8,
        }
      }
    >
      Đọc ít nhất 30 phút mỗi ngày
    </Text>
    <Text
      style={
        {
          "color": "#999",
          "fontSize": 12,
          "fontWeight": "500",
        }
      }
    >
      Streak: 
      5
       ngày
    </Text>
  </View>
  <View
    accessibilityRole="checkbox"
    accessibilityState={
      {
        "busy": undefined,
        "checked": false,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "alignItems": "center",
        "backgroundColor": "#F9F9F9",
        "borderColor": "#E5E5EA",
        "borderRadius": 20,
        "borderWidth": 2,
        "height": 40,
        "justifyContent": "center",
        "opacity": 1,
        "width": 40,
      }
    }
    testID="habit-card-toggle"
  >
    <Text
      style={
        [
          {
            "color": "#999",
            "fontSize": 20,
          },
          false,
        ]
      }
    >
      ○
    </Text>
  </View>
</View>
`;
